pam/integration-tests/sshd_preloader/sshd_preloader.c

```diff
@@ -0,0 +1,133 @@
+#define _GNU_SOURCE 1
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <dlfcn.h>
+#include <string.h>
+#include <pwd.h>
+#include <limits.h>
+
+#define AUTHD_TEST_SHELL "/bin/sh"
+#define AUTHD_TEST_GECOS ""
+#define AUTHD_DEFAULT_SSH_PAM_SERVICE_NAME "sshd"
+#define AUTHD_SPECIAL_USER_ACCEPT_ALL "authd-test-user-sshd-accept-all"
+
+static struct passwd passwd_entities[512];
+
+__attribute__((constructor))
+void constructor (void)
+{
+  fprintf (stderr, "sshd_preloader: Library loaded (parent pid :%d)!\n",
+           getpid ());
+}
+
+__attribute__((destructor))
+void destructor (void)
+{
+  fprintf (stderr, "sshd_preloader: Library unloaded  (parent pid :%d)!\n",
+           getpid ());
+}
+
+static const char *
+get_home_path (void)
+{
+  const char *home_path = getenv ("AUTHD_TEST_SSH_HOME");
+
+  if (home_path == NULL)
+    return "/not-existing-home";
+
+  return home_path;
+}
+
+static bool
+test_user_is_accepted (const char *test_user,
+                       const char *name)
+{
+  if (test_user == NULL || *test_user == '\0')
+    return false;
+
+  if (strcmp (test_user, name) == 0)
+    return true;
+
+  if (strcmp (test_user, AUTHD_SPECIAL_USER_ACCEPT_ALL) != 0)
+    return false;
+
+  /* Here we accept all the users supported by the example broker */
+  if (strncmp (name, "user", 4) == 0)
+    return true;
+
+  /* Further special case for the 'r' user */
+  if (strcmp (name, "r") == 0)
+    return true;
+
+  return false;
+}
+
+struct passwd *
+getpwnam (const char *name)
```

Comment:
Could you add a comment here which explains why we hook `getpwnam` instead of letting authd handle the NSS request?